File: src/api/routers/compression_config.py
"""
Compression Config API Router

Provides endpoints for configuring context compression.
"""
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field
from typing import Literal, Optional
import logging

from ..services.compression_config_service import CompressionConfigService

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/compression", tags=["compression"])
ALLOWED_OUTPUT_LANGUAGES = {"auto", "none", "zh", "en", "ja", "ko", "fr", "de", "es", "ru", "pt"}


# Pydantic models
class CompressionConfigResponse(BaseModel):
    """Response model for compression configuration"""
    provider: str
    model_id: str
    local_gguf_model_path: str
    local_gguf_n_ctx: int
    local_gguf_n_threads: int
    local_gguf_n_gpu_layers: int
    local_gguf_max_tokens: int
    temperature: float
    min_messages: int
    timeout_seconds: int
    prompt_template: str
    compression_output_language: str
    compression_strategy: str
    hierarchical_chunk_target_tokens: int
    hierarchical_chunk_overlap_messages: int
    hierarchical_reduce_target_tokens: int
    hierarchical_reduce_overlap_items: int
    hierarchical_max_levels: int
    quality_guard_enabled: bool
    quality_guard_min_coverage: float
    quality_guard_max_facts: int
    compression_metrics_enabled: bool
    auto_compress_enabled: bool
    auto_compress_threshold: float


class CompressionConfigUpdate(BaseModel):
    """Request model for updating compression configuration"""
    provider: Optional[str] = None
    model_id: Optional[str] = None
    local_gguf_model_path: Optional[str] = None
    local_gguf_n_ctx: Optional[int] = Field(None, ge=512, le=65536)
    local_gguf_n_threads: Optional[int] = Field(None, ge=0, le=256)
    local_gguf_n_gpu_layers: Optional[int] = Field(None, ge=0, le=1024)
    local_gguf_max_tokens: Optional[int] = Field(None, ge=64, le=16384)
    temperature: Optional[float] = Field(None, ge=0.0, le=2.0)
    min_messages: Optional[int] = Field(None, ge=1, le=50)
    timeout_seconds: Optional[int] = Field(None, ge=10, le=300)
    prompt_template: Optional[str] = None
    compression_output_language: Optional[str] = None
    compression_strategy: Optional[Literal["single_pass", "hierarchical"]] = None
    hierarchical_chunk_target_tokens: Optional[int] = Field(None, ge=0, le=8192)
    hierarchical_chunk_overlap_messages: Optional[int] = Field(None, ge=0, le=20)
    hierarchical_reduce_target_tokens: Optional[int] = Field(None, ge=0, le=16384)
    hierarchical_reduce_overlap_items: Optional[int] = Field(None, ge=0, le=10)
    hierarchical_max_levels: Optional[int] = Field(None, ge=1, le=8)
    quality_guard_enabled: Optional[bool] = None
    quality_guard_min_coverage: Optional[float] = Field(None, ge=0.5, le=1.0)
    quality_guard_max_facts: Optional[int] = Field(None, ge=5, le=100)
    compression_metrics_enabled: Optional[bool] = None
    auto_compress_enabled: Optional[bool] = None
    auto_compress_threshold: Optional[float] = Field(None, ge=0.1, le=0.9)


# Dependency
def get_compression_config_service() -> CompressionConfigService:
    """Get CompressionConfigService instance"""
    return CompressionConfigService()


# Endpoints
@router.get("/config", response_model=CompressionConfigResponse)
async def get_config(
    service: CompressionConfigService = Depends(get_compression_config_service)
):
    """Get current compression configuration"""
    try:
        config = service.config
        return CompressionConfigResponse(
            provider=config.provider,
            model_id=config.model_id,
            local_gguf_model_path=config.local_gguf_model_path,
            local_gguf_n_ctx=config.local_gguf_n_ctx,
            local_gguf_n_threads=config.local_gguf_n_threads,
            local_gguf_n_gpu_layers=config.local_gguf_n_gpu_layers,
            local_gguf_max_tokens=config.local_gguf_max_tokens,
            temperature=config.temperature,
            min_messages=config.min_messages,
            timeout_seconds=config.timeout_seconds,
            prompt_template=config.prompt_template,
            compression_output_language=config.compression_output_language,
            compression_strategy=config.compression_strategy,
            hierarchical_chunk_target_tokens=config.hierarchical_chunk_target_tokens,
            hierarchical_chunk_overlap_messages=config.hierarchical_chunk_overlap_messages,
            hierarchical_reduce_target_tokens=config.hierarchical_reduce_target_tokens,
            hierarchical_reduce_overlap_items=config.hierarchical_reduce_overlap_items,
            hierarchical_max_levels=config.hierarchical_max_levels,
            quality_guard_enabled=config.quality_guard_enabled,
            quality_guard_min_coverage=config.quality_guard_min_coverage,
            quality_guard_max_facts=config.quality_guard_max_facts,
            compression_metrics_enabled=config.compression_metrics_enabled,
            auto_compress_enabled=config.auto_compress_enabled,
            auto_compress_threshold=config.auto_compress_threshold,
        )
    except Exception as e:
        logger.error(f"Failed to get compression config: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/config")
async def update_config(
    updates: CompressionConfigUpdate,
    service: CompressionConfigService = Depends(get_compression_config_service)
):
    """Update compression configuration"""
    try:
        update_dict = updates.model_dump(exclude_none=True)

        if not update_dict:
            raise HTTPException(status_code=400, detail="No updates provided")

        if 'provider' in update_dict:
            allowed = {"model_config", "local_gguf"}
            if update_dict['provider'] not in allowed:
                raise HTTPException(
                    status_code=400,
                    detail=f"Unsupported compression provider: {update_dict['provider']}"
                )
        if 'compression_output_language' in update_dict:
            value = str(update_dict['compression_output_language']).strip().lower()
            if value not in ALLOWED_OUTPUT_LANGUAGES:
                raise HTTPException(
                    status_code=400,
                    detail=f"Unsupported compression output language: {update_dict['compression_output_language']}"
                )
            update_dict['compression_output_language'] = value
        if 'compression_strategy' in update_dict:
            allowed_strategies = {"single_pass", "hierarchical"}
            if update_dict['compression_strategy'] not in allowed_strategies:
                raise HTTPException(
                    status_code=400,
                    detail=f"Unsupported compression strategy: {update_dict['compression_strategy']}"
                )

        service.save_config(update_dict)

        return {"message": "Configuration updated successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to update compression config: {e}")
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/modules/settings/config/compression.config.ts
/**
 * Compression Settings Configuration
 *
 * Defines the structure and behavior of the Context Compression settings page.
 */

import type { SimpleConfigSettingsConfig } from './types';
import i18n from '../../../i18n';

export const compressionConfig: SimpleConfigSettingsConfig = {
  type: 'config',
  get title() { return i18n.t('settings:compression.title'); },
  get description() { return i18n.t('settings:compression.description'); },

  apiEndpoint: {
    get: '/api/compression/config',
    update: '/api/compression/config'
  },

  fields: [
    {
      type: 'select',
      name: 'provider',
      get label() { return i18n.t('settings:compression.field.provider'); },
      required: true,
      defaultValue: 'model_config',
      get options() {
        return [
          { value: 'model_config', label: i18n.t('settings:compression.opt.modelConfig') },
          { value: 'local_gguf', label: i18n.t('settings:compression.opt.localGguf') },
        ];
      },
      get helpText() { return i18n.t('settings:compression.field.provider.help'); }
    },
    {
      type: 'checkbox',
      name: 'auto_compress_enabled',
      get label() { return i18n.t('settings:compression.field.autoCompress'); },
      defaultValue: false
    },
    {
      type: 'select',
      name: 'compression_output_language',
      get label() { return i18n.t('settings:compression.field.outputLanguage'); },
      defaultValue: 'auto',
      options: [
        { value: 'auto', label: i18n.t('settings:compression.opt.outputLanguageAuto') },
        { value: 'none', label: i18n.t('settings:compression.opt.outputLanguageNone') },
        { value: 'zh', label: i18n.t('settings:compression.opt.outputLanguageZh') },
        { value: 'en', label: i18n.t('settings:compression.opt.outputLanguageEn') },
        { value: 'ja', label: i18n.t('settings:compression.opt.outputLanguageJa') },
        { value: 'ko', label: i18n.t('settings:compression.opt.outputLanguageKo') },
        { value: 'fr', label: i18n.t('settings:compression.opt.outputLanguageFr') },
        { value: 'de', label: i18n.t('settings:compression.opt.outputLanguageDe') },
        { value: 'es', label: i18n.t('settings:compression.opt.outputLanguageEs') },
        { value: 'ru', label: i18n.t('settings:compression.opt.outputLanguageRu') },
        { value: 'pt', label: i18n.t('settings:compression.opt.outputLanguagePt') },
      ],
      get helpText() { return i18n.t('settings:compression.field.outputLanguage.help'); },
    },
    {
      type: 'slider',
      name: 'auto_compress_threshold',
      get label() { return i18n.t('settings:compression.field.threshold'); },
      min: 0.1,
      max: 0.9,
      step: 0.05,
      defaultValue: 0.5,
      get helpText() { return i18n.t('settings:compression.field.threshold.help'); },
      condition: (formData) => formData.auto_compress_enabled === true
    },
    {
      type: 'select',
      name: 'model_id',
      get label() { return i18n.t('settings:compression.field.modelId'); },
      required: true,
      dynamicOptions: (context) => {
        const models = context.models || [];
        const modelOptions = models.map((m: any) => ({
          value: `${m.provider_id}:${m.id}`,
          label: m.name || m.id
        }));
        return [
          { value: 'same_as_chat', label: i18n.t('settings:compression.opt.sameAsChatModel') },
          ...modelOptions,
        ];
      },
      get helpText() { return i18n.t('settings:compression.field.modelId.help'); },
      condition: (formData) => formData.provider !== 'local_gguf',
    },
    {
      type: 'text',
      name: 'local_gguf_model_path',
      get label() { return i18n.t('settings:compression.field.localGgufModelPath'); },
      get placeholder() { return i18n.t('settings:compression.field.localGgufModelPath.placeholder'); },
      defaultValue: 'models/llm/local-summarizer.gguf',
      get helpText() { return i18n.t('settings:compression.field.localGgufModelPath.help'); },
      condition: (formData) => formData.provider === 'local_gguf',
    },
    {
      type: 'number',
      name: 'local_gguf_n_ctx',
      get label() { return i18n.t('settings:compression.field.localGgufCtx'); },
      min: 512,
      max: 65536,
      defaultValue: 8192,
      get helpText() { return i18n.t('settings:compression.field.localGgufCtx.help'); },
      condition: (formData) => formData.provider === 'local_gguf',
    },
    {
      type: 'number',
      name: 'local_gguf_n_threads',
      get label() { return i18n.t('settings:compression.field.localGgufThreads'); },
      min: 0,
      max: 256,
      defaultValue: 0,
      get helpText() { return i18n.t('settings:compression.field.localGgufThreads.help'); },
      condition: (formData) => formData.provider === 'local_gguf',
    },
    {
      type: 'number',
      name: 'local_gguf_n_gpu_layers',
      get label() { return i18n.t('settings:compression.field.localGgufGpuLayers'); },
      min: 0,
      max: 1024,
      defaultValue: 0,
      get helpText() { return i18n.t('settings:compression.field.localGgufGpuLayers.help'); },
      condition: (formData) => formData.provider === 'local_gguf',
    },
    {
      type: 'number',
      name: 'local_gguf_max_tokens',
      get label() { return i18n.t('settings:compression.field.localGgufMaxTokens'); },
      min: 64,
      max: 16384,
      defaultValue: 2048,
      get helpText() { return i18n.t('settings:compression.field.localGgufMaxTokens.help'); },
      condition: (formData) => formData.provider === 'local_gguf',
    },
    {
      type: 'select',
      name: 'compression_strategy',
      get label() { return i18n.t('settings:compression.field.strategy'); },
      defaultValue: 'hierarchical',
      options: [
        { value: 'hierarchical', label: i18n.t('settings:compression.opt.hierarchical') },
        { value: 'single_pass', label: i18n.t('settings:compression.opt.singlePass') },
      ],
      get helpText() { return i18n.t('settings:compression.field.strategy.help'); },
      condition: (formData) => formData.provider === 'local_gguf',
    },
    {
      type: 'number',
      name: 'hierarchical_chunk_target_tokens',
      get label() { return i18n.t('settings:compression.field.hierChunkTargetTokens'); },
      min: 0,
      max: 8192,
      defaultValue: 0,
      get helpText() { return i18n.t('settings:compression.field.hierChunkTargetTokens.help'); },
      condition: (formData) => formData.provider === 'local_gguf' && formData.compression_strategy === 'hierarchical',
    },
    {
      type: 'number',
      name: 'hierarchical_chunk_overlap_messages',
      get label() { return i18n.t('settings:compression.field.hierChunkOverlapMessages'); },
      min: 0,
      max: 20,
      defaultValue: 2,
      get helpText() { return i18n.t('settings:compression.field.hierChunkOverlapMessages.help'); },
      condition: (formData) => formData.provider === 'local_gguf' && formData.compression_strategy === 'hierarchical',
    },
    {
      type: 'number',
      name: 'hierarchical_reduce_target_tokens',
      get label() { return i18n.t('settings:compression.field.hierReduceTargetTokens'); },
      min: 0,
      max: 16384,
      defaultValue: 0,
      get helpText() { return i18n.t('settings:compression.field.hierReduceTargetTokens.help'); },
      condition: (formData) => formData.provider === 'local_gguf' && formData.compression_strategy === 'hierarchical',
    },
    {
      type: 'number',
      name: 'hierarchical_reduce_overlap_items',
      get label() { return i18n.t('settings:compression.field.hierReduceOverlapItems'); },
      min: 0,
      max: 10,
      defaultValue: 1,
      get helpText() { return i18n.t('settings:compression.field.hierReduceOverlapItems.help'); },
      condition: (formData) => formData.provider === 'local_gguf' && formData.compression_strategy === 'hierarchical',
    },
    {
      type: 'number',
      name: 'hierarchical_max_levels',
      get label() { return i18n.t('settings:compression.field.hierMaxLevels'); },
      min: 1,
      max: 8,
      defaultValue: 4,
      get helpText() { return i18n.t('settings:compression.field.hierMaxLevels.help'); },
      condition: (formData) => formData.provider === 'local_gguf' && formData.compression_strategy === 'hierarchical',
    },
    {
      type: 'checkbox',
      name: 'quality_guard_enabled',
      get label() { return i18n.t('settings:compression.field.qualityGuardEnabled'); },
      defaultValue: true,
      condition: (formData) => formData.provider === 'local_gguf',
    },
    {
      type: 'slider',
      name: 'quality_guard_min_coverage',
      get label() { return i18n.t('settings:compression.field.qualityGuardMinCoverage'); },
      min: 0.5,
      max: 1.0,
      step: 0.05,
      defaultValue: 0.75,
      get helpText() { return i18n.t('settings:compression.field.qualityGuardMinCoverage.help'); },
      condition: (formData) => formData.provider === 'local_gguf' && formData.quality_guard_enabled !== false,
    },
    {
      type: 'number',
      name: 'quality_guard_max_facts',
      get label() { return i18n.t('settings:compression.field.qualityGuardMaxFacts'); },
      min: 5,
      max: 100,
      defaultValue: 24,
      get helpText() { return i18n.t('settings:compression.field.qualityGuardMaxFacts.help'); },
      condition: (formData) => formData.provider === 'local_gguf' && formData.quality_guard_enabled !== false,
    },
    {
      type: 'checkbox',
      name: 'compression_metrics_enabled',
      get label() { return i18n.t('settings:compression.field.metricsEnabled'); },
      defaultValue: true,
      get helpText() { return i18n.t('settings:compression.field.metricsEnabled.help'); },
    },
    {
      type: 'slider',
      name: 'temperature',
      get label() { return i18n.t('settings:compression.field.temperature'); },
      min: 0,
      max: 2,
      step: 0.1,
      defaultValue: 0.3,
      get helpText() { return i18n.t('settings:compression.field.temperature.help'); }
    },
    {
      type: 'number',
      name: 'min_messages',
      get label() { return i18n.t('settings:compression.field.minMessages'); },
      min: 1,
      max: 50,
      defaultValue: 2,
      required: true,
      get helpText() { return i18n.t('settings:compression.field.minMessages.help'); }
    },
    {
      type: 'number',
      name: 'timeout_seconds',
      get label() { return i18n.t('settings:compression.field.timeout'); },
      min: 10,
      max: 300,
      defaultValue: 60,
      required: true,
      get helpText() { return i18n.t('settings:compression.field.timeout.help'); }
    },
    {
      type: 'textarea',
      name: 'prompt_template',
      get label() { return i18n.t('settings:compression.field.promptTemplate'); },
      rows: 10,
      monospace: true,
      get placeholder() { return i18n.t('settings:compression.field.promptTemplate.placeholder'); },
      get helpText() { return i18n.t('settings:compression.field.promptTemplate.help'); }
    }
  ]
};
